File: infrastructure/integration-test/configuration_test/configuration_test.f90
!-----------------------------------------------------------------------------
! (C) Crown copyright 2017 Met Office. All rights reserved.
! The file LICENCE, distributed with this code, contains details of the terms
! under which the code may be used.
!-----------------------------------------------------------------------------

program configuration_test

  use ESMF,                        only : ESMF_Finalize,   &
                                          ESMF_Initialize, &
                                          ESMF_VM,         &
                                          ESMF_VMGet
  use iso_fortran_env,             only : error_unit, output_unit
  use one_of_each_test_config_mod, only : key_from_an_enum,                  &
                                      postprocess_one_of_each_test_namelist, &
                                          read_one_of_each_test_namelist,    &
                                          angle_deg,                         &
                                          angle_rad,                         &
                                          an_enum,                           &
                                          closed_array,                      &
                                          max_array_size,                    &
                                          open_array,                        &
                                          some_string,                       &
                                          whole_number

  implicit none

  integer,      parameter :: file_unit = 13
  character(*), parameter :: filename = 'one_of_each.nml'

  type(ESMF_VM) :: vm
  integer       :: rank
  integer       :: condition
  character(30) :: format_string

  call ESMF_Initialize( vm=vm, rc=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to initialise ESMF")' )
    stop 1
  end if

  call ESMF_VMGet( vm, localPET=rank, rc=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to get rank from ESMF")' )
    stop 2
  end if

  open( file_unit, file=filename, iostat=condition )

  if (condition /= 0) then
    write( error_unit, '("Failed to open file: ",A)' ) filename
    stop 3
  end if

  call read_one_of_each_test_namelist( file_unit, vm ,rank )

  close( file_unit, iostat=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to close file: ", A)' ) filename
    stop 4
  end if

  call postprocess_one_of_each_test_namelist

  call ESMF_Finalize( rc=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to finalise ESMF")' )
    stop 5
  end if

  write( output_unit, '("angle_deg: ", E14.7)' ) angle_deg
  write( output_unit, '("angle_rad: ", E14.7)' ) angle_rad
  write( output_unit, '("an_enum: ", A)' ) key_from_an_enum( an_enum )
  write( format_string, &
         '("(""closed_array: "", ", I0, "E14.7)")' ) max_array_size
  write( output_unit, format_string ) closed_array
  write( format_string, '("(""open_array: "", ", I0, "I3)")' ) max_array_size
  write( output_unit, format_string ) open_array
  write( output_unit, '("some_string: ''", A, "''")' ) trim(some_string)
  write( output_unit, '("whole_number: ", I0)' ) whole_number

end program configuration_test

File: infrastructure/integration-test/configuration_test/configuration_test.py
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-
##############################################################################
# (c) Crown copyright 2017 Met Office. All rights reserved.
# The file LICENCE, distributed with this code, contains details of the terms
# under which the code may be used.
##############################################################################

from __future__ import print_function

import re
import sys

from testframework import Test, TestEngine, TestFailed

##############################################################################
class one_of_each_test(Test):
  def __init__( self ):
    self._INJECT = 'one_of_each.nml'
    super(one_of_each_test, self).__init__( [sys.argv[1], self._INJECT] )

    self._precision = 0.0005

  def test( self, process ):
    out, err = process.communicate( )
    if process.returncode != 0:
      raise TestFailed( 'Unexpected failure of test executable: {code}' \
                        .format( code=process.returncode ) )

    expected = { 'angle_deg'   : 7.998,
                 'angle_rad'   : 0.1396,
                 'an_enum'     : 'second',
                 'closed_array': [0.2, 0.3, 0.4],
                 'open_array'  : [1, 2, 3, 4, 5],
                 'some_string' : 'chocolate teapot',
                 'whole_number': 13}
    seen = []
    variablePattern = re.compile( r'(.+?)\s*:\s*(.+)\s*' )
    listPattern = re.compile( r'(\'.*?\'|".*?"|[^ ]+)' )
    for line in out.splitlines():
      match = variablePattern.match( line )
      if match:
        variable = match.group(1).strip()
        seen.append( variable )
        value    = match.group(2).strip()

        listMatch = listPattern.findall( value )
        if isinstance(listMatch, list):
          value = []
          for item in listMatch:
            if item.startswith("'") or item.startswith('"'):
              value.append( item[1:-1] )
            else:
              value.append( item )

        if variable not in expected:
          message = 'Found unexpected variable "{variable}" in output'
          raise TestFailed( message.format( variable=variable ) )

        self._check( variable, expected[variable], value )

    additions = set(seen) - set(expected.keys())
    if len(additions) > 0:
      message = 'Addition variables found: {adds}'
      raise TestFailed( message.format( adds=', '.join( additions ) ) )

    return 'One of each configuration type loaded'

  def _check( self, name, expected, found ):
    if not isinstance( expected, list ):
      expected = [expected]

    if not isinstance( found, list ):
      found = [found]

    if len(found) != len(expected):
      message = 'Expected variable "{variable}" to hold "{expect}" but found "{found}", different number of values.'
      raise TestFailed( message.format( variable=name,
                                        expect=expected,
                                        found=found ) )

    for index in range( len(expected) ):
      if isinstance( expected[index], float):
        if abs(float(found[index]) - expected[index]) > self._precision:
          message = 'Expected index {index} of variable "{variable}" to be within {precision} of {expected} but found {found}'
          raise TestFailed( message.format( variable=name,
                                            index=index,
                                            precision=self._precision,
                                            expected=expected[index],
                                            found=found[index] ) )
      else:
        if found[index] != str(expected[index]):
          message = 'Expected index {index} of variable "{variable}" to hold "{expect}" but found "{found}"'
          raise TestFailed( message.format( variable=name,
                                            index=index,
                                            expect=expected[index],
                                            found=found[index] ) )

##############################################################################
if __name__ == '__main__':
  TestEngine.run( one_of_each_test() )
